File: perl5/update_readme_stats.pl
#!/usr/bin/perl
# update_readme_stats.py — portiert nach perl5
# Quelle: python, OpenClaw@main:scripts/update_readme_stats.py
# Erzeugt: 2026-08-22 durch ABSTRACTIONS_MANAGER.py

use strict;
use warnings;
use utf8;
use LWP::UserAgent;
use JSON qw(decode_json);
use Encode qw(decode);

# Fetch ClawHub stats and update README.md download counts and security status.

my $API_BASE = "https://clawhub.ai/api/v1";
my $TOKEN = $ENV{"CLAWHUB_TOKEN"} // "";

my @SKILLS = (
    ["Cluster Gateway",           "cluster-gateway"],
    ["MCP Tool Utils",            "mcp-tool-utils"],
    ["Reports Creator",           "reports-creator"],
    ["Relay Node",                "relay-node"],
    ["JSON Utils",                "json-utils"],
    ["Log Collector",             "log-collector"],
    ["TikTok Live Monitor",       "tiktok-live-monitor"],
    ["Doc Scraper",               "doc-scraper"],
    ["Workspace Database Manager","workspace-database-manager"],
    ["Scripting Utils",           "scripting-utils"],
);

sub fetch_skill {
    my ($slug) = @_;
    my $url = "$API_BASE/skills/$slug";
    my $ua = LWP::UserAgent->new;
    $ua->timeout(10);
    my $req = HTTP::Request->new(GET => $url);
    if ($TOKEN) {
        $req->header("Authorization" => "Bearer $TOKEN");
    }
    $req->header("Accept" => "application/json");
    my $res = $ua->request($req);
    if (!$res->is_success) {
        die "HTTP " . $res->code . ": " . $res->message;
    }
    return decode_json($res->content);
}

sub parse_skill {
    my ($data) = @_;
    my $skill   = $data->{"skill"} // {};
    my $stats   = $skill->{"stats"} // {};
    my $version_data = $data->{"latestVersion"} // {};
    my $version = $version_data->{"version"} // "1.0.0";
    my $mod     = $data->{"moderation"};

    my $downloads = $stats->{"downloads"} // 0;

    my $security;
    if (!defined $mod) {
        $security = "✅ Pass";
    } elsif ($mod->{"isMalwareBlocked"}) {
        $security = "🚫 Blocked";
    } else {
        $security = "🔍 Review";
    }

    return {
        "downloads" => $downloads,
        "version"   => $version =~ /^v/ ? $version : "v$version",
        "security"  => $security,
    };
}

sub main {
    my %stats = ();
    my $errors = 0;
    foreach my $entry (@SKILLS) {
        my ($name, $slug) = @$entry;
        eval {
            my $data = fetch_skill($slug);
            my $s = parse_skill($data);
            $stats{$slug} = $s;
            print "  OK  $slug: $s->{\"downloads\"} downloads, $s->{\"version\"}, $s->{\"security\"}\n";
        };
        if ($@) {
            print STDERR "  ERR $slug: $@\n";
            $errors++;
        }
    }

    if (keys %stats == 0) {
        print STDERR "No data fetched — aborting.\n";
        exit 1;
    }

    my $content;
    {
        open(my $fh, "<:encoding(UTF-8)", "README.md") or die "Cannot read README.md: $!";
        local $/;
        $content = <$fh>;
        close($fh);
    }

    foreach my $entry (@SKILLS) {
        my ($name, $slug) = @$entry;
        if (!exists $stats{$slug}) {
            next;
        }
        my $dl = $stats{$slug}->{"downloads"};
        # Escape special regex characters in the name
        my $escaped_name = quotemeta($name);
        my $pattern = qr/(\|\s*\[?$escaped_name\]?[^|]*\|[^|]*\|)\s*\d+\s*(\|)/i;
        my $replacement = "\$1 $dl \$2";
        my $new_content = $content;
        $new_content =~ s/$pattern/$replacement/g;
        if ($new_content ne $content) {
            $content = $new_content;
            print "  Updated: $name -> $dl\n";
        }
    }

    open(my $fh, ">:encoding(UTF-8)", "README.md") or die "Cannot write README.md: $!";
    print $fh $content;
    close($fh);
    print "Done: " . (scalar keys %stats) . " skills, $errors errors.\n";
}

main() unless caller;

File: javascript/update_readme_stats.js
#!/usr/bin/env node
// update_readme_stats.py — portiert nach javascript
// Quelle: python, OpenClaw@main:scripts/update_readme_stats.py
// Erzeugt: 2026-08-22 durch ABSTRACTIONS_MANAGER.py

/** Fetch ClawHub stats and update README.md download counts and security status. */

const fs = require('fs');
const https = require('https');

const API_BASE = "https://clawhub.ai/api/v1";
const TOKEN = process.env.CLAWHUB_TOKEN || "";

const SKILLS = [
    ["Cluster Gateway",           "cluster-gateway"],
    ["MCP Tool Utils",            "mcp-tool-utils"],
    ["Reports Creator",           "reports-creator"],
    ["Relay Node",                "relay-node"],
    ["JSON Utils",                "json-utils"],
    ["Log Collector",             "log-collector"],
    ["TikTok Live Monitor",       "tiktok-live-monitor"],
    ["Doc Scraper",               "doc-scraper"],
    ["Workspace Database Manager","workspace-database-manager"],
    ["Scripting Utils",           "scripting-utils"],
];

function fetchSkill(slug) {
    return new Promise((resolve, reject) => {
        const url = `${API_BASE}/skills/${slug}`;
        const options = {
            headers: {
                'Accept': 'application/json'
            }
        };
        if (TOKEN) {
            options.headers.Authorization = `Bearer ${TOKEN}`;
        }
        
        https.get(url, options, (res) => {
            let data = '';
            res.on('data', (chunk) => {
                data += chunk;
            });
            res.on('end', () => {
                try {
                    resolve(JSON.parse(data));
                } catch (err) {
                    reject(err);
                }
            });
        }).on('error', (err) => {
            reject(err);
        }).setTimeout(10000, () => {
            reject(new Error('Request timeout'));
        });
    });
}

function parseSkill(data) {
    const skill = data.skill || {};
    const stats = skill.stats || {};
    const versionData = data.latestVersion || {};
    const version = versionData.version || "1.0.0";
    const mod = data.moderation;

    const downloads = stats.downloads || 0;

    let security = "✅ Pass";
    if (mod === null || mod === undefined) {
        security = "✅ Pass";
    } else if (mod.isMalwareBlocked) {
        security = "🚫 Blocked";
    } else {
        security = "🔍 Review";
    }

    return {
        downloads: downloads,
        version: version.startsWith("v") ? version : `v${version}`,
        security: security,
    };
}

async function main() {
    const stats = {};
    let errors = 0;
    
    for (const [name, slug] of SKILLS) {
        try {
            const data = await fetchSkill(slug);
            const s = parseSkill(data);
            stats[slug] = s;
            console.log(`  OK  ${slug}: ${s.downloads} downloads, ${s.version}, ${s.security}`);
        } catch (exc) {
            console.error(`  ERR ${slug}: ${exc}`);
            errors++;
        }
    }

    if (Object.keys(stats).length === 0) {
        console.error("No data fetched — aborting.");
        process.exit(1);
    }

    let content = fs.readFileSync("README.md", "utf-8");

    for (const [name, slug] of SKILLS) {
        if (!(slug in stats)) {
            continue;
        }
        const dl = stats[slug].downloads;
        const escapedName = name.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
        const pattern = new RegExp(`(\\|\\s*\\[?${escapedName}\\]?[^|]*\\|[^|]*\\|)\\s*\\d+\\s*(\\|)`, 'i');
        const replacement = `$1 ${dl} $2`;
        const newContent = content.replace(pattern, replacement);
        if (newContent !== content) {
            content = newContent;
            console.log(`  Updated: ${name} -> ${dl}`);
        }
    }

    fs.writeFileSync("README.md", content, "utf-8");
    console.log(`Done: ${Object.keys(stats).length} skills, ${errors} errors.`);
}

main().catch(err => {
    console.error(err);
    process.exit(1);
});
